docs/kb/wasm.rst

```diff
@@ -0,0 +1,307 @@
+            (call $fib (i64.sub (local.get $n) (i64.const 1)))
+            (call $fib (i64.sub (local.get $n) (i64.const 2)))
+            )
+        )
+        (memory (;0;) 2)
+        (export "memory" (memory 0))
+        (export "fib" (func $fib))
+        (global (;0;) i32 (i32.const 1024))
+        (export "_scylla_abi" (global 0))
+        (data $.rodata (i32.const 1024) "\01")
+    )'
+
+and it can be invoked just like a regular UDF:
+
+.. code-block:: cql
+
+    scylla@cqlsh:ks> CREATE TABLE t(id int, n bigint, PRIMARY KEY(id,n));
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 0);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 1);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 2);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 3);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 4);
+    scylla@cqlsh:ks> INSERT INTO t(id, n) VALUES (0, 5);
+    scylla@cqlsh:ks> SELECT n, ks.fib(n) FROM t;
+
+    n  | ks.fib(n)
+    ----+-----------
+     0 |         0
+     1 |         1
+     2 |         1
+     3 |         2
+     4 |         3
+     5 |         5
+
+    (11 rows)
+
+Experimental status
+-------------------
+
+WebAssembly UDFs are still experimental due to insufficient testing. If backwards incompatible changes
+to the ABI are implemented in the future, they should be submitted as new ABI-versions, and use the same
+``LANGUAGE wasm`` clause in the CQL statements.
+
+ABI versions
+------------
+
+Different programming languages may require different ABIs. To support that, the Wasm program is required to
+export the symbol ``_scylla_abi``, that is a WebAssembly global with a 32-bit value of the offset in memory,
+where the version number can be read (that's the only method of exporting a constant in Rust).
+
+Currently, the only available ABI versions are 1 and 2. Both of them use the same protocol for passing
+parameters and returning values, but they differ in approaches to memory management.
+
+Memory management
+-----------------
+
+The memory management differs depending on the used ABI verison:
+
+* version 1
+    There are no requirements of the usage of memory by the user. The host grows memory for each of parameters and does not free the memory in any way.
+
+* version 2
+    The user program is required to export ``_scylla_malloc`` and ``_scylla_free`` methods, which
+    are then used by the host for allocating memory for parameters and freeing memory for the returned value.
+    The user is required to free the memory allocated for parameters using the ``_scylla_free`` method, and allocate the memory for result using the ``_scylla_malloc`` method (both can be achieved by using the provided helper libraries).
+    Alternatively, the user may return one of the arguments, shifting the responsibility of freeing it to the host.
+    The ``_scylla_malloc`` and ``_scylla_free`` methods may be simple wrappers of ``malloc`` and ``free`` methods implemented by default when compiling with WASI.
```

Comment:
Do these functions even need to be defined? Can't they just be "extern"? If it's just extern it will be clearer that the definition doesn't matter - it's Scylla that implements these functions, no?

Reply:
So I guess I just don't understand how this works. If _scylla_malloc() is defined as calling malloc(), where does this malloc() come from, from Scylla and its shared libraries? Can the user call **any** function whatsoever from UDF? Isn't this a security problem? For example, the user can create a call to system("rm -rf /")? It doesn't make sense we want to allow that.

Reply:
This malloc appears in the Wasm module when compiling to the wasm32-wasi target (in fact, it is often a big part of the entire module when compiling simple functions). You can actually even see it here: https://github.com/scylladb/scylladb/blob/8ceb8b0240721ea1bf938a39f28eb49b34e01233/test/resource/wasm/return_input.wat#L14-L17
It doesn't use system calls, instead it operates only on the Wasm Memory. We can export some functions from the host to the Wasm module, but we have to do it explicitly and currently we're not exporting any.

Reply:
Wow, now I understand.. So, it seems the library (is it your library or clang's?) contains an implementation of `malloc()` which does the right (and safe) thing. 
So, why do we want to force the user to define `_scylla_malloc` to call `malloc()`? Does it ever make sense to define it to anything else? If we know the library must implement `malloc()`, why doesn't Scylla just call that instead of `scylla_malloc`?

Reply:
We decided on it quite some time ago already, but I remember that the first reason was that there were problems with exporting the generated `malloc` from Wasm, and the other one is to let the user (or ourselves in a helper library) use their own allocator (compiling to wasm32-wasi is not strictly required by us, but it's recommended due to the presence of this `malloc`)

Reply:
Ok. I understood none of that while reading the document, but maybe I only cared about this as a developer, and not a "user" who would just use the instructions and not stop to think about every little detail why it has to be like that.

Reply:
Yes, not all intentions were described in the doc, but even if they did, I guess we would leave them in the docs/dev. And as you mentioned, we do expect users to keep using using the Rust helper library which hides all the internals away from the user